x-pack/plugins/security/public/session/session_timeout.ts

```diff
@@ -0,0 +1,342 @@
+ * redirect to logout portal etc.).
+ */
+export const GRACE_PERIOD_MS = 5 * 1000;
+
+/**
+ * Duration we'll normally display the warning toast
+ */
+export const WARNING_MS = 5 * 60 * 1000;
+
+/**
+ * Current session info is checked this number of milliseconds before the warning toast shows. This
+ * will prevent the toast from being shown if the session has already been extended.
+ */
+export const SESSION_CHECK_MS = 1000;
+
+/**
+ * Session will be extended at most once this number of milliseconds while user activity is detected.
+ */
+export const EXTENSION_THROTTLE_MS = 60 * 1000;
+
+/**
+ * Route to get session info and extend session expiration
+ */
+export const SESSION_ROUTE = '/internal/security/session';
+
+export interface SessionState extends Pick<SessionInfo, 'expiresInMs' | 'canBeExtended'> {
+  lastExtensionTime: number;
+}
+
+export class SessionTimeout {
+  private channel?: BroadcastChannel<SessionState>;
+
+  private isVisible = document.visibilityState !== 'hidden';
+  private isExtending = false;
+
+  private sessionState$ = new BehaviorSubject<SessionState>({
+    lastExtensionTime: 0,
+    expiresInMs: null,
+    canBeExtended: false,
+  });
+  private subscription?: Subscription;
+
+  private warningToast?: Toast;
+
+  private stopActivityMonitor?: Function;
+  private stopVisibilityMonitor?: Function;
+  private removeHttpInterceptor?: Function;
+
+  private stopRefreshTimer?: Function;
+  private stopWarningTimer?: Function;
+  private stopLogoutTimer?: Function;
+
+  constructor(
+    private notifications: NotificationsSetup,
+    private sessionExpired: ISessionExpired,
+    private http: HttpSetup,
+    private tenant: string
+  ) {}
+
+  public start() {
+    if (this.http.anonymousPaths.isAnonymous(window.location.pathname)) {
+      return;
+    }
+
+    this.subscription = this.sessionState$
+      .pipe(skip(1), throttleTime(1000), tap(this.toggleEventHandlers))
+      .subscribe(this.resetTimers);
+
+    return this.fetchSessionInfo();
+  }
+
+  public stop() {
+    const nextState = {
+      lastExtensionTime: 0,
+      expiresInMs: null,
+      canBeExtended: false,
+    };
+    this.toggleEventHandlers(nextState);
+    this.resetTimers(nextState);
+    this.subscription?.unsubscribe();
+  }
+
+  /**
+   * Event handler that receives session information from other browser tabs.
+   */
+  private handleChannelMessage = (message: SessionState) => {
+    this.sessionState$.next(message);
+  };
+
+  /**
+   * HTTP request interceptor which ensures that API calls extend the session only if tab is
+   * visible.
+   */
+  private handleHttpRequest = (fetchOptions: HttpFetchOptionsWithPath) => {
+    // Ignore requests to external URLs
+    if (fetchOptions.path.indexOf('://') !== -1) {
+      return;
+    }
+
+    if (fetchOptions.asSystemRequest === undefined) {
+      return { asSystemRequest: !this.isVisible };
+    }
```

Comment:
It looks like this will clobber other fetch options. Also if some other code has explicitly set `asSystemRequest` to false, I think it's safe to say we want to override it here.

```suggestion
    if (!fetchOptions.asSystemRequest && !this.isVisible) {
      return { ...fetchOptions, asSystemRequest: true };
    }
```

Reply:
Ah, good catch, that would have been a nightmare to debug! 

If `asSystemRequest` is explicitly set to false already then we don't need to do anything so it's better to treat it as a noop in that case and return void.

Reply:
> If `asSystemRequest` is explicitly set to false already then we don't need to do anything so it's better to treat it as a noop in that case and return void.

I'm not sure I follow? Regardless of what the consumer has explicitly, I think if the page is not visible, we want to set `asSystemRequest` to true, right? Otherwise the session would be extended when the user is not using the page.

Reply:
My assumption would be that we don't want to do that. If a developer explicitly sets `asSystemRequest` to false I would trust that they've got a good reason to do so.

Reply:
I feel pretty strongly that we don't want other plugins to circumvent security controls. There is no reason to explicitly set `asSystemRequest` to false, if a developer does that, they are misunderstanding the option, and it shouldn't result in an infinitely extended session when this is used for polling requests.